### PR TITLE
Track active requests in S3FileSystemExchangeStorage

### PR DESCRIPTION
**Pull request by arhimondr**
## Description

To help diagnose the issue described in https://github.com/trinodb/trino/issues/14496

## Non-technical explanation

N/A

## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java [hunk 1]
```diff
@@ -163,12 +171,41 @@ public S3FileSystemExchangeStorage(S3FileSystemExchangeStorageStats stats, Excha
...
+                responseFuture.whenComplete((result, failure) -> {
+                    requestsActive.get(requestType).decrementAndGet();
+                    logActiveRequestsIfNecessary();
+                    if (failure != null && failure.getMessage() != null && failure.getMessage().contains("Maximum pending connection acquisitions exceeded")) {
+                        log.error("Encountered the 'Maximum pending connection acquisitions exceeded' error. Requests active: %s", requestsActive);
+                    }
+                });
+            }
+
+            private void logActiveRequestsIfNecessary()
```
Comment (martint) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java:200]: Can’t this be exposed via JMX instead?

Comment (arhimondr) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java]: Yeah, I think JMX is a much better option. I added simple logging to unblock debugging while working on JMX export. Implemented through JMX now.

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3AsyncClientWrapper.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    @Override
+    public ListObjectsV2Publisher listObjectsV2Paginator(ListObjectsV2Request listObjectsV2Request)
+    {
+        return new ListObjectsV2Publisher(this, applyPaginatorUserAgent(listObjectsV2Request));
+    }
+
+    /**
+     * Based on {@link software.amazon.awssdk.services.s3.DefaultS3AsyncClient#applyPaginatorUserAgent(S3Request)}
+     */
+    private <T extends S3Request> T applyPaginatorUserAgent(T request)
```
Comment (losipiuk) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3AsyncClientWrapper.java:145]: where exactly are the separate requests for pages handled?

Comment (arhimondr) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3AsyncClientWrapper.java:145]: This is done by passing the reference to the wrapper to the `ListObjectsV2Publisher`

> return new ListObjectsV2Publisher(this, applyPaginatorUserAgent(listObjectsV2Request));

**Comment by linzebing** [comment]
We no longer need this PR, right?

**Comment by losipiuk** [comment]
> We no longer need this PR, right?

Exposing stats via JMX is still nice. I would keep it.

**Review by losipiuk — APPROVED** [review]

**Review by linzebing — APPROVED** [review]

**Pull request merged**